### PR TITLE
Add option to disable ambiguous unicode characters detection

### LINKED ISSUE #23149: Incorrect report of ambiguous unicode characters
### Description

The wiki page viewer (and editor) show a warring on ambiguous unicode chararcters I tried for most non-latin1 characters. Please remove it at all. It is hardly useful even would it be functional (and it is not). It is not possible to make a good guess on which unicode is OK for a project with many localisations. The locale of the developer (reporter, wiki writer) computer/browser often will be different from the locale that is used to report a problem (in a localised text for example) or describe a test case described in the wiki page.

### Gitea Version

1.18.5

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

Screenshot 2023-02-25 at 15 06 10

### Git Version

1.18.5

### Operating System

Linux ** 4.19.0-21-amd64 #1 SMP Debian 4.19.249-2 (2022-06-30) x86_64

### How are you running Gitea?

downloaded your binaries

### Database

PostgreSQL

### LINKED ISSUE #23682: Wrong unicode char reporting.
image

### LINKED ISSUE #24483: Add configuration option to disable check of "This file contains ambiguous Unicode characters!"
### Feature Description

There should be a configuration option to disable check of "This file contains ambiguous Unicode characters!"
Our project has different language comments and it triggers this banner to be displayed on every file. Whant to have a possibility to disable it

### Screenshots

chars

### LINKED ISSUE #28123: Implement option to disable globally "This file contains ambiguous Unicode characters!" annoying message.
### Feature Description

In my installation I use russian language in README.md files and other places such as comments in src.

And this thing annoy me and my team. I'd like to have ini-config flag or setting in web-ui to disable this "feature" permanently for all instance of gitea/all repos.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by wxiaoguang**
* Close #24483
* Close #28123
* Close #23682
* Close #23149

(maybe more)

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Comment by silverwind** [comment]
Maybe it should even be disabled by default because as I see it, the current detection is almost pure false-positives for the "warnings", only the "errors" are real problems.

**Comment by lunny** [comment]
> Maybe it should even be disabled by default because as I see it, the current detection is almost pure false-positives for the "warnings", only the "errors" are real problems.

Then we need a break label.

**Comment by silverwind** [comment]
> > Maybe it should even be disabled by default because as I see it, the current detection is almost pure false-positives for the "warnings", only the "errors" are real problems.
> 
> Then we need a break label.

It wouldn't break anyone's workflow I assume. Not a fan of unnecessary breaking labels.

**Comment by wxiaoguang** [comment]
> Maybe it should even be disabled by default because as I see it, the current detection is almost pure false-positives for the "warnings", only the "errors" are real problems.

I won't do so at the moment, because otherwise some extremists would catch this point and criticize that "Gitea doesn't take security seriously“

**Comment by wxiaoguang** [comment]
> I think ideally we just remove the warning category (which warns on any irregular whitespace), and just keep the error category (which are the real bidi exploits). Doesn't have to be in this PR.

That's why I used `AMBIGUOUS_UNICODE_DETECTION = true`, in the future it could be improved to : `AMBIGUOUS_UNICODE_DETECTION = error` or `AMBIGUOUS_UNICODE_DETECTION = bidi`, etc

**Comment by JakobDev** [comment]
> I think ideally we just remove the warning category

Sounds like a good idea. Currently this feature is kinda useless. If you have almost only false positives, Users do not care about the Warning and will ignore it, even when there's a real threat.

You should also add a User setting. Not everyone is Admin.

**Comment by wxiaoguang** [comment]
> > I think ideally we just remove the warning category
> 
> Sounds like a good idea. Currently this feature is kinda useless. If you have almost only false positives, Users do not care about the Warning and will ignore it, even when there's a real threat.
> 
> You should also add a User setting. Not everyone is Admin.

It doesn't work that way at the moment, every detected problem is rendered as "warning".

Any improvement could be done later.

image

**Comment by silverwind** [comment]
Does this hide the "Escape" button in diffs? I think it should when the feature is disabled.

**Comment by wxiaoguang** [comment]
> Does this hide the "Escape" button in file view and diffs? I think it should when the feature is disabled as the button is only useful for that feature.

Yes. There won't be "escape" button if there is no detected ambiguous unicode character.

File: routers/web/repo/view.go [hunk 1]
```diff
@@ -317,19 +319,18 @@ func renderReadmeFile(ctx *context.Context, subfolder string, readmeFile *git.Tr
...
-		buf := &bytes.Buffer{}
-		ctx.Data["EscapeStatus"], err = charset.EscapeControlStringReader(rd, buf, ctx.Locale)
+		content, err := io.ReadAll(rd)
 		if err != nil {
-			log.Error("Read failed: %v", err)
+			log.Error("Read readme content failed: %v", err)
 		}
-
-		ctx.Data["FileContent"] = buf.String()
+		contentEscaped := template.HTMLEscapeString(unsafe.String(&content[0], len(content)))
```
Comment (wolfogre) [inline routers/web/repo/view.go]: Do we need to check if the length of the content is not zero? Or has the preceding code already ensured that the content cannot be empty?

Comment (wolfogre) [reply routers/web/repo/view.go]: Maybe
```suggestion
		contentEscaped := template.HTMLEscapeString(unsafe.String(unsafe.SliceData(content), len(content)))
```

Comment (wxiaoguang) [reply routers/web/repo/view.go]: Good catch, I think we can also do this 848533d57b7cd79e11c79dd438fc6e8f937bf020

**Review by wolfogre — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> > Does this hide the "Escape" button in file view and diffs? I think it should when the feature is disabled as the button is only useful for that feature.
> 
> Yes. There won't be "escape" button if there is no detected ambiguous unicode character.

Just realized that some "escape" buttons are hidden correctly, while some are not, because some of the old code doesn't respect `EscapeStatus.Escaped` .... so some buttons are always there.

**Comment by silverwind** [comment]
> > > Does this hide the "Escape" button in file view and diffs? I think it should when the feature is disabled as the button is only useful for that feature.
> > 
> > 
> > Yes. There won't be "escape" button if there is no detected ambiguous unicode character.
> 
> Just realized that some "escape" buttons are hidden correctly, while some are not, because some of the old code doesn't respect `EscapeStatus.Escaped` .... so some buttons are always there.

We should get rid of all those unneccesary and confusing escape buttons, e.g. only show if there is actually escapable content.

**Pull request merged**